### PR TITLE
apps: change list command to only list fetchable algorithms

### LINKED ISSUE #15196: Filtering openssl list output by provider
When I call 
`openssl list -provider-path `pwd`/providers/ -provider legacy -propquery "provider=legacy" -digest-algorithms -kdf-algorithms -mac-algorithms -cipher-algorithms -key-exchange-algorithms -kem-algorithms -signature-algorithms -asymcipher-algorithms -public-key-algorithms` 
it lists all the algorithms including definitely not legacy, such as Ed448. 

Is it an intended behavior?

### PR DESCRIPTION
**Pull request by paulidale**
The `-propquery` option will work with this change.  By default the output will be the same.

Other inconsistencies in the error checking are cleaned up.

Fixes #15196

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: apps/list.c [hunk 1]
```diff
@@ -115,8 +112,12 @@ static void list_ciphers(void)
         BIO_printf(bio_err, "ERROR: Memory allocation\n");
         return;
     }
-    BIO_printf(bio_out, "Legacy:\n");
-    EVP_CIPHER_do_all_sorted(legacy_cipher_fn, bio_out);
+#ifndef OPENSSL_NO_DEPRECATED_3_0
+    if (include_legacy) {
+        BIO_printf(bio_out, "Legacy:\n");
```
Comment (levitte) [inline apps/list.c:117]: Maybe it should say "Legacy implementations", to be super clear

Comment (levitte) [reply apps/list.c:117]: (not just in this spot)

Comment (paulidale) [reply apps/list.c:117]: I'd prefer to not use the term "legacy" at all here.  It's confusing.
How about "compatibility implementations" instead?

Comment (paulidale) [reply apps/list.c:117]: The "Legacy" label is what was there BTW :grinning:

Comment (paulidale) [reply apps/list.c:117]: I'm thinking "Compatibility" might be better.  It sits more nicely with "Provided" which also doesn't have the "implementations".  That or add "implementations" to both (which doesn't seem necessary).

Comment (paulidale) [reply apps/list.c]: And done.  I added a `-backwards-compatibility` option to enable the display of legacy implementations in listings.

File: doc/man1/openssl-list.pod.in [hunk 2]
```diff
@@ -10,6 +10,9 @@ openssl-list - list algorithms and features
 B<openssl list>
 [B<-help>]
 [B<-verbose>]
+{- output_off() if $disabled{"deprecated-3.0"}; ""
+-}[B<-backwards-compatibility>]
+{- output_on() if $disabled{"deprecated-3.0"}; ""
 [B<-select> I<name>]
```
Comment (levitte) [inline doc/man1/openssl-list.pod.in:16]: You forgot a closing brace

File: apps/list.c [hunk 3]
```diff
@@ -26,9 +26,42 @@
...
+IS_FETCHABLE(rand, EVP_RAND)
+IS_FETCHABLE(keymgmt, EVP_KEYMGMT)
+IS_FETCHABLE(signature, EVP_SIGNATURE)
+IS_FETCHABLE(kem, EVP_KEM)
+IS_FETCHABLE(asym_cipher, EVP_ASYM_CIPHER)
+IS_FETCHABLE(keyexch, EVP_KEYEXCH)
+IS_FETCHABLE(decoder, OSSL_DECODER)
+IS_FETCHABLE(encoder, OSSL_ENCODER)
+
+#ifndef OPENSSL_NO_DEPRECATED_3_0
```
Comment (levitte) [inline apps/list.c:64]: You should wrap `list_md_fn` in the same manner

File: CHANGES.md [hunk 4]
```diff
@@ -23,10 +23,16 @@ OpenSSL 3.0
 
 ### Changes between 1.1.1 and 3.0 [xx XXX xxxx]
 
+ * The list command will only display provider based implementations by default.
+   The -backwards-compatibility must be specified to display algorithm
+   implementations that exist for legacy and engine purposes.
```
Comment (beldmit) [inline CHANGES.md:28]: Well... That's not the best behavior, I think. 

If no `propquery` is provided, I think, the `list` command should enumerate all the available algorithms. 
If `propquery` is provided, iy should be applied (and skip all the engine-provided algorithms).

File: doc/man1/openssl-list.pod.in (hunk 2 above)
Comment (paulidale) [reply doc/man1/openssl-list.pod.in:16]: I'll just undo this in favour of @beldmit's suggestion which seems better all around.

File: apps/list.c (hunk 3 above)
Comment (paulidale) [reply apps/list.c:64]: Good point. I was looking for a function `legacy_digest_fn` :(
I'll rename it at the same time.

File: CHANGES.md (hunk 4 above)
Comment (paulidale) [reply CHANGES.md]: And done.

**Comment by paulidale** [comment]
Feedback addressed.  Out of WIP.

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by beldmit** [comment]
Thanks @paulidale, nice work!

**Comment by levitte** [comment]
So, this isn't draft any more, I guess?

**Comment by paulidale** [comment]
Not draft and I forget to include the final changes.

File: doc/man3/OSSL_ENCODER.pod [hunk 5]
```diff
@@ -74,6 +76,8 @@ algorithm that's identifiable with I<name>.
 OSSL_ENCODER_number() returns the internal dynamic number assigned to
 the given I<encoder>.
 
+OSSL_ENCODER_number() returns the name used to fetch the given I<encoder>.
```
Comment (t8m) [inline doc/man3/OSSL_ENCODER.pod:79]: Copy & paste error. Also it is not the name used to fetch but the first name from the provider name list. You could just copy the description from for example EVP_SIGNATURE_name().

File: doc/man3/OSSL_DECODER.pod [hunk 6]
```diff
@@ -74,6 +76,8 @@ of an algorithm that's identifiable with I<name>.
 OSSL_DECODER_number() returns the internal dynamic number assigned
 to the given I<decoder>.
 
+OSSL_DECODER_number() returns the name used to fetch the given I<decoder>.
```
Comment (t8m) [inline doc/man3/OSSL_DECODER.pod:79]: Same as below for OSSL_ENCODER_name().

File: doc/man3/OSSL_ENCODER.pod (hunk 5 above)
Comment (paulidale) [reply doc/man3/OSSL_ENCODER.pod:79]: Yeah, should have looked elsewhere for sample text.  Didn't think of it.

File: crypto/encode_decode/decoder_meth.c [hunk 7]
```diff
@@ -425,6 +430,11 @@ int OSSL_DECODER_number(const OSSL_DECODER *decoder)
     return decoder->base.id;
 }
 
+const char *OSSL_DECODER_name(const OSSL_DECODER *decoder)
```
Comment (t-j-h) [inline crypto/encode_decode/decoder_meth.c:433]: OSSL_DECODER_get_name()

Comment (paulidale) [reply crypto/encode_decode/decoder_meth.c:433]: **All** of the other functions are `XXX_name()`...

**Comment by t-j-h** [comment]
The OSSL_DECODER and OSSL_ENCODER getters aren't really named the same way we have named everything else.
You are following the pattern here - and I think they should all be changed - but not in this PR unless you want to tackle that here.
Perhaps a separate PR?

**Comment by paulidale** [comment]
Rebased and squashed to see if checksums needed updating.  No functional changes.

**Comment by paulidale** [comment]
Merged to master. Thanks for the feedback.

**Pull request closed without merging**